### DISCUSSION TITLE
How to upload a database with extenssion .bak to asset library of LCS?

### DISCUSSION THREAD

**Discussion by celuxcwb** [d365collaborative/d365fo.tools#614] [Q&A] [answered]
First of all, I would like to thank the project founders and the community for their efforts in maintaining a sensational project like this one, which my team at the company where I work use every day and which we hope to contribute with some new features in the future. next.

We recently implemented a pipeline on our Build machine to export the Production data overnight and in this way import bacpac to a local SQL Server and generate a backup with *.Bak extension, all steps were successful, however using a Soccer metaphor "we are hitting the post with the ball" when uploading the final file (.bak) to the LCS asset library.

In an old version of the documentation we find the comand: 
[https://www.powershellgallery.com/packages/d365fo.tools/0.5.4/Content/functions%5Cinvoke-d365lcsupload.ps1](url)

`.EXAMPLE
        PS C:\> Invoke-D365LcsUpload -ProjectId 123456789 -ClientId "9b4f4503-b970-4ade-abc6-2c086e4c4929" -Username claire@contoso.com -Password "pass@word1" -FilePath "C:\temp\d365fo.tools\GOLDEN.bacpac" -FileType "DatabaseBackup" -FileName "ReadyForTesting" -FileDescription "Contains all customers & vendors" -LcsApiUri "https://lcsapi.lcs.dynamics.com"`
        
however this command is no longer supported in current versions of D365FO.Tools, which is a shame, however there is a reason or limitation for this.

Is there an alternative way for us to meet our goal of uploading the Backup to the LCS using the tool? If not, what would these limitations be?

Thanks in advance for any help or support from the community.

I apologize if my question is not very clear or my translation is not understandable.

**Comment by FH-Inway** [comment]
Maybe a misunderstanding, but as far as I can see, the cmdlet is still available. Check [Invoke-D365LcsUpload.md](https://github.com/d365collaborative/d365fo.tools/blob/development/docs/Invoke-D365LcsUpload.md) for the documentation and [Invoke-D365LcsUpload.ps1](https://github.com/d365collaborative/d365fo.tools/blob/development/d365fo.tools/functions/invoke-d365lcsupload.ps1) for the cmdlet itself.

That said, for an upload in a pipeline (assuming Azure DevOps pipeline), you may want to consider the [Upload assets](https://docs.microsoft.com/en-us/dynamics365/fin-ops-core/dev-itpro/dev-tools/pipeline-asset-upload) pipeline task instead, which is part of the [Dynamics 365 Finance and Operations Tools](https://marketplace.visualstudio.com/items?itemName=Dyn365FinOps.dynamics365-finops-tools) extension for Azure DevOps.

Btw, your question reads very clear 👍

**Comment by celuxcwb** [comment]
Thanks @FH-Inway  for the answer.

After the end of the process, we tried to proceed with the creation of a release pipeline using the file upload task (as we do with our deployable packages), however the file upload task does not offer the option of Database Backups, no were it for this limitation this would be a great way to achieve the goal.

![FileOptions](https://user-images.githubusercontent.com/13489789/149154610-5f8682f6-2f1a-4abc-bcca-2e8febce5f31.PNG)

**Comment by Splaxi** [comment, marked as answer]
Thanks for taking the time to post the question 👍

The issue is that Microsoft **very clearly** doesn't want us to upload bak file by automation directly into LCS. The reasons why are a bit unclear, but I believe it might be that some customers / partners used LCS as a backup location for bak files - which isn't what Microsoft intended to offer. So that is why the removed the support for bak file upload, and why you can see that we supported it in an earlier version of the module.

With that said - why is it important for you to have the bak file uploaded to LCS?

**Comment by celuxcwb** [comment]
Thanks for the answer @Splaxi .

The final objective would be to create a routine that developers could run on their machines and update their local databases, which would generate some time savings in relation to this restore routine.

But I understand that Microsoft may have some restrictions on how customers are using this feature.

Eventually I believe that using a storage account to store backups can be a way out.

**Reply by Splaxi** [reply]
We do exactly that for all our customers. 

From the bacpac is exported to the asset library in lcs, we move it to our own storage account. When the bak file is prepped, we push it to the storage account. 

The developer execute a local script and gets a fresh copy in less than 30 minutes.

**Reply by Jnagasuresh** [reply]
Hello @Splaxi 


Can you please help me with below issue. Please let me know if any additional details are needed.

I am trying to upload the .bacpac file from my local drive to LCS asster library, and I am end with an error which does not have much details. Please help if anybody have solutions. Please let me know if any further questions.

Here is the error I am receiving on console:
[17:15:56][Start-LcsUploadV2] Something went wrong while working against the LCS API. | Response status code does not indicate success: 500 (Internal Server Error).
WARNING: [17:15:56][Start-LcsUploadV2] Stopping because of errors

Source code;
$LcsApiUri = "https://lcsapi.lcs.dynamics.com/"
[int]$ProjectId = 1666824
($TokenResponse.access_token)" -- Getting token it right...
$LocalPath = "D:\bacpac\sivatestbackup.bacpac"
#$FileType = "SoftwareDeployablePackage"
$FileType = "ProcessDataPackage"
try
{

#Get-PSFMessage -Level Verbose | Write-Host
Invoke-D365LcsUpload -ProjectId $ProjectId `
 -BearerToken $BearerToken `
 -FilePath $LocalPath `
 -LcsApiUri $LcsApiUri `
 -FileType $FileType `
 -FileName "ReadyForTesting" `
 -FileDescription "testing file upload" `
 -RetryTimeout "02:00:30" 
#Get-PSFMessage -Level Verbose | Write-Host
#(Get-D365EnvironmentSetting).Aos
}
catch {
$formattedDateTime = Get-Date -Format "yyyy-MM-dd HH:mm:ss"
Write-Host "Error occurred while uploading file to asset library tatus: $_ at: current time = $formattedDateTime"
return "Failed"
}

**Reply by FH-Inway** [reply]
@Jnagasuresh You also created issue #788 with that problem description. Note that it may very well be that Microsoft not only blocks uploads of .bak files, but uploads of all backup files.

**Comment by Jnagasuresh** [comment]
HI All,

is there a solution for above problem now?

I am facing same issue. Thanks in advance for your help.

**Reply by FH-Inway** [reply]
As mentioned in the answer by Splaxi, Microsoft blocks uploads of .bak files to the asset library. I'm not aware that this has changed and the fact that you are facing the issue seems to indicate further that the block is still in place.

As discussed, you can use a different storage such as Azure storage account instead.
